ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,573 @@
+        break;
+      case Disk_Error:
+        ((AmbryDisk) replica.getDiskId()).onDiskError();
+        break;
+      case Disk_Ok:
+        ((AmbryDisk) replica.getDiskId()).onDiskOk();
+        break;
+      case Partition_ReadOnly:
+        ((AmbryPartition) replica.getPartitionId()).onPartitionReadOnly();
+        break;
+    }
+  }
+
+  @Override
+  public PartitionId getPartitionIdFromStream(InputStream stream) throws IOException {
+    byte[] partitionBytes = AmbryPartition.readPartitionBytesFromStream(stream);
+    AmbryPartition partition = partitionMap.get(ByteBuffer.wrap(partitionBytes));
+    if (partition == null) {
+      throw new IOException("Partition id from stream is null");
+    }
+    return partition;
+  }
+
+  /**
+   * @return list of partition ids that are in {@link PartitionState#READ_WRITE}.
+   */
+  @Override
+  public List<PartitionId> getWritablePartitionIds() {
+    List<PartitionId> writablePartitions = new ArrayList<>();
+    List<PartitionId> healthyWritablePartitions = new ArrayList<>();
+    for (AmbryPartition partition : partitionNameToPartitionIdMap.values()) {
+      if (partition.getPartitionState() == PartitionState.READ_WRITE) {
+        writablePartitions.add(partition);
+        if (areAllReplicasForPartitionUp(partition)) {
+          healthyWritablePartitions.add(partition);
+        }
+      }
+    }
+    return healthyWritablePartitions.isEmpty() ? writablePartitions : healthyWritablePartitions;
+  }
+
+  /**
+   * Check whether all replicas of the given {@link AmbryPartition} are up.
+   * @param partition the {@link AmbryPartition} to check.
+   * @return true if all associated replicas are up; false otherwise.
+   */
+  private boolean areAllReplicasForPartitionUp(AmbryPartition partition) {
+    for (AmbryReplica replica : partitionIdToReplicaIds.get(partition)) {
+      if (replica.isDown()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Return the unique {@link AmbryReplica} for a {@link AmbryPartition} on a {@link AmbryDataNode}.
+   * @param hostname the hostname associated with the {@link AmbryDataNode}.
+   * @param port the port associated with the {@link AmbryDataNode}.
+   * @param partitionString the partition id string associated with the {@link AmbryPartition}.
+   * @return the {@link AmbryReplica} associated with the given parameters.
+   */
+  AmbryReplica getReplicaForPartitionOnNode(String hostname, int port, String partitionString) {
+    for (AmbryReplica replica : dataNodeIdToReplicaIds.get(getDataNodeId(hostname, port))) {
+      if (replica.getPartitionId().toString().equals(partitionString)) {
+        return replica;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * An instance of this object is used to register as listener for Helix related changes in each datacenter.
+   */
+  private class ClusterChangeListener implements ExternalViewChangeListener, ConfigChangeListener, LiveInstanceChangeListener {
```

Comment:
Yes - have replaced it with `InstanceConfigChangeListener` now.
(previously had to use the deprecated one, because `RoutingTableProvider` still used it and for some reason registering as an `InstanceConfigChangeListener` did not work out - no notifications on changes).